File: Scrapper/JavaScrapper/scrapper/src/main/java/scrapper.java
import org.jsoup.Jsoup;
import org.jsoup.nodes.Document;
import org.jsoup.nodes.Element;
import org.jsoup.select.Elements;

import java.io.FileOutputStream;
import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.URL;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

public class scrapper {

    private static final String BASE_URL = "https://papers.nips.cc";
    private static final String SAVE_DIRECTORY = "E:\\pdfs\\";
    private static final ExecutorService executor = Executors.newFixedThreadPool(10); // Adjust pool size as needed

    public static void main(String[] args) {
        try {
            Path savePath = Paths.get(SAVE_DIRECTORY);
            if (!Files.exists(savePath)) {
                Files.createDirectories(savePath);
                System.out.println("Created save directory: " + savePath);
            }

            System.out.println("Fetching main page: " + BASE_URL);
            Document mainPage = Jsoup.connect(BASE_URL).get();

            Elements yearLinks = mainPage.select("a[href^=/paper_files/paper/]");
            System.out.println("Found " + yearLinks.size() + " year links.");

            for (Element yearLink : yearLinks) {
                String yearUrl = BASE_URL + yearLink.attr("href");

                // Extract the year from the link text (e.g., "2024" from the link)
                String yearName = yearUrl.substring(yearUrl.lastIndexOf('/') + 1);
                Path yearFolderPath = Paths.get(SAVE_DIRECTORY, yearName);

                // Create the year folder if it doesn't exist
                if (!Files.exists(yearFolderPath)) {
                    Files.createDirectories(yearFolderPath);
                    System.out.println("Created year folder: " + yearFolderPath);
                }

                System.out.println("Processing year: " + yearUrl);
                Document yearPage = Jsoup.connect(yearUrl).get();

                Elements paperLinks = yearPage.select("a[href^=/paper_files/paper/]");

                for (Element paperLink : paperLinks) {
                    String paperPageUrl = BASE_URL + paperLink.attr("href");
                    System.out.println("Processing paper page: " + paperPageUrl);

                    Document paperPage = Jsoup.connect(paperPageUrl).get();

                    Element downloadButton = paperPage.selectFirst("a[href$=.pdf]");
                    if (downloadButton != null) {
                        String pdfUrl = BASE_URL + downloadButton.attr("href");
                        System.out.println("Downloading paper: " + pdfUrl);

                        executor.submit(() -> downloadFile(pdfUrl, yearFolderPath.toString()));
                    } else {
                        System.err.println("Download button not found for paper: " + paperPageUrl);
                    }
                }
            }
        } catch (IOException e) {
            e.printStackTrace();
        } finally {
            executor.shutdown();
        }
    }

    private static void downloadFile(String fileUrl, String saveDirectory) {
        try {
            URL url = new URL(fileUrl);
            HttpURLConnection connection = (HttpURLConnection) url.openConnection();
            connection.setRequestMethod("GET");

            String fileName = fileUrl.substring(fileUrl.lastIndexOf('/') + 1);
            Path filePath = Paths.get(saveDirectory, fileName);

            if (!Files.exists(filePath.getParent())) {
                Files.createDirectories(filePath.getParent());
            }

            try (InputStream in = connection.getInputStream();
                 FileOutputStream out = new FileOutputStream(filePath.toFile())) {
                byte[] buffer = new byte[4096];
                int bytesRead;
                while ((bytesRead = in.read(buffer)) != -1) {
                    out.write(buffer, 0, bytesRead);
                }
            }

            System.out.println("Saved: " + filePath);
        } catch (IOException e) {
            System.err.println("Failed to download: " + fileUrl);
            e.printStackTrace();
        }
    }
}

File: Scrapper/PythonScrapper/Scrapper.py
import os
import requests
from bs4 import BeautifulSoup
from concurrent.futures import ThreadPoolExecutor

BASE_URL = "https://papers.nips.cc"
SAVE_DIRECTORY = "E:\\pdfs\\"
THREAD_POOL_SIZE = 10
TIMEOUT = 10  # Timeout for requests in seconds


def create_directory(path):
    if not os.path.exists(path):
        os.makedirs(path)
        print(f"Created directory: {path}")


def fetch_main_page(url):
    print(f"Fetching main page: {url}")
    response = requests.get(url, timeout=TIMEOUT)
    response.raise_for_status()
    return BeautifulSoup(response.text, 'html.parser')


def fetch_year_page(url):
    print(f"Processing year page: {url}")
    response = requests.get(url, timeout=TIMEOUT)
    response.raise_for_status()
    return BeautifulSoup(response.text, 'html.parser')


def fetch_paper_page(url):
    print(f"Processing paper page: {url}")
    response = requests.get(url, timeout=TIMEOUT)
    response.raise_for_status()
    return BeautifulSoup(response.text, 'html.parser')


def download_file(pdf_url, save_directory):
    try:
        file_name = pdf_url.split('/')[-1]
        file_path = os.path.join(save_directory, file_name)

        response = requests.get(pdf_url, stream=True, timeout=TIMEOUT)
        response.raise_for_status()

        with open(file_path, 'wb') as file:
            for chunk in response.iter_content(chunk_size=4096):
                file.write(chunk)

        print(f"Saved: {file_path}")
    except requests.RequestException as e:
        print(f"Failed to download: {pdf_url}\nError: {e}")


if __name__ == "__main__":
    create_directory(SAVE_DIRECTORY)

    try:
        main_page = fetch_main_page(BASE_URL)

        year_links = main_page.select("a[href^=/paper_files/paper/]")
        print(f"Found {len(year_links)} year links.")

        with ThreadPoolExecutor(max_workers=THREAD_POOL_SIZE) as executor:
            for year_link in year_links:
                year_url = BASE_URL + year_link['href']

                year_name = year_url.split('/')[-1]
                year_folder_path = os.path.join(SAVE_DIRECTORY, year_name)
                create_directory(year_folder_path)

                year_page = fetch_year_page(year_url)
                paper_links = year_page.select("a[href^=/paper_files/paper/]")

                for paper_link in paper_links:
                    paper_page_url = BASE_URL + paper_link['href']
                    paper_page = fetch_paper_page(paper_page_url)
                    
                    download_button = paper_page.select_one("a[href$=.pdf]")
                    if download_button:
                        pdf_url = BASE_URL + download_button['href']
                        print(f"Downloading paper: {pdf_url}")
                        executor.submit(download_file, pdf_url, year_folder_path)
                    else:
                        print(f"Download button not found for paper: {paper_page_url}")
    except requests.exceptions.RequestException as e:
        print(f"Error fetching main page or processing: {e}")
